pkg/controller/machinedeployment/util/util.go

```diff
@@ -0,0 +1,693 @@
+	// Then, update machine set's revision annotation
+	if newMS.Annotations == nil {
+		newMS.Annotations = make(map[string]string)
+	}
+	oldRevision, ok := newMS.Annotations[RevisionAnnotation]
+	// The newMS's revision should be the greatest among all MSes. Usually, its revision number is newRevision (the max revision number
+	// of all old MSes + 1). However, it's possible that some of the old MSes are deleted after the newMS revision being updated, and
+	// newRevision becomes smaller than newMS's revision. We should only update newMS revision when it's smaller than newRevision.
+
+	oldRevisionInt, err := strconv.ParseInt(oldRevision, 10, 64)
+	if err != nil {
+		if oldRevision != "" {
+			glog.Warningf("Updating machine set revision OldRevision not int %s", err)
+			return false
+		}
+		//If the MS annotation is empty then initialise it to 0
+		oldRevisionInt = 0
+	}
+	newRevisionInt, err := strconv.ParseInt(newRevision, 10, 64)
+	if err != nil {
+		glog.Warningf("Updating machine set revision NewRevision not int %s", err)
+		return false
+	}
+	if oldRevisionInt < newRevisionInt {
+		newMS.Annotations[RevisionAnnotation] = newRevision
+		annotationChanged = true
+		glog.V(4).Infof("Updating machine set %q revision to %s", newMS.Name, newRevision)
+	}
+	// If a revision annotation already existed and this machine set was updated with a new revision
+	// then that means we are rolling back to this machine set. We need to preserve the old revisions
+	// for historical information.
+	if ok && annotationChanged {
+		revisionHistoryAnnotation := newMS.Annotations[RevisionHistoryAnnotation]
+		oldRevisions := strings.Split(revisionHistoryAnnotation, ",")
+		if len(oldRevisions[0]) == 0 {
+			newMS.Annotations[RevisionHistoryAnnotation] = oldRevision
+		} else {
+			oldRevisions = append(oldRevisions, oldRevision)
+			newMS.Annotations[RevisionHistoryAnnotation] = strings.Join(oldRevisions, ",")
+		}
+	}
+	// If the new machine set is about to be created, we need to add replica annotations to it.
+	if !exists && SetReplicasAnnotations(newMS, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+MaxSurge(*deployment)) {
+		annotationChanged = true
+	}
+	return annotationChanged
+}
+
+// FindActiveOrLatest returns the only active or the latest machine set in case there is at most one active
+// machine set. If there are more active machine sets, then we should proportionally scale them.
+func FindActiveOrLatest(newMS *v1alpha1.MachineSet, oldMSs []*v1alpha1.MachineSet) *v1alpha1.MachineSet {
+	if newMS == nil && len(oldMSs) == 0 {
+		return nil
+	}
+
+	sort.Sort(sort.Reverse(MachineSetsByCreationTimestamp(oldMSs)))
+	allMSs := FilterActiveMachineSets(append(oldMSs, newMS))
+
+	switch len(allMSs) {
+	case 0:
+		// If there is no active machine set then we should return the newest.
+		if newMS != nil {
+			return newMS
+		}
+		return oldMSs[0]
+	case 1:
+		return allMSs[0]
+	default:
+		return nil
```

Comment:
Is this behaviour intended?  It doesn't really match the comment I think - allMSs[0] is the latest I think

Reply:
The naming was off. It really wanted one active or latest. If there are multiple machine sets that are available, it needs to start scaling down some of the older machine sets first to avoid having too many active machine sets.

renamed the method and updated the comment.